File: scripts/siteline_list_type_fields.py
import argparse
import os
import sys
from typing import Any, Dict, List

import requests


SCHEMA_QUERY = """
query __TypeFields($name: String!) {
  __type(name: $name) {
    fields {
      name
    }
  }
}
"""


def load_env(path: str) -> Dict[str, str]:
    """
    Minimal .env loader (KEY=VALUE). This avoids adding new dependencies.
    """
    env: Dict[str, str] = {}
    if not os.path.exists(path):
        return env

    with open(path, "r", encoding="utf-8") as f:
        for raw in f:
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            if "=" not in line:
                continue
            k, v = line.split("=", 1)
            k = k.strip()
            v = v.strip().strip("'").strip('"')
            env[k] = v
    return env


def get_headers(api_token: str, auth_header: str) -> Dict[str, str]:
    if auth_header:
        return {auth_header: api_token}
    return {"Authorization": f"Bearer {api_token}"}


def list_fields(api_url: str, api_token: str, auth_header: str, gql_type_name: str) -> List[str]:
    headers = get_headers(api_token, auth_header)
    resp = requests.post(
        api_url,
        json={"query": SCHEMA_QUERY, "variables": {"name": gql_type_name}},
        headers={**headers, "Content-Type": "application/json"},
        timeout=60,
    )
    resp.raise_for_status()

    payload: Dict[str, Any] = resp.json()
    if "errors" in payload and payload["errors"]:
        raise RuntimeError(f"GraphQL errors: {payload['errors']}")

    t = payload.get("data", {}).get("__type")
    if not t:
        return []
    fields = t.get("fields") or []
    return sorted([f["name"] for f in fields if f.get("name")])


def main() -> None:
    parser = argparse.ArgumentParser(
        description="List GraphQL field names for Siteline types using introspection (__type(fields))."
    )
    parser.add_argument(
        "--type",
        dest="types",
        action="append",
        required=True,
        help="GraphQL type name to introspect (repeatable). Example: --type Project --type Contract",
    )
    parser.add_argument("--env-file", default=os.path.join(os.getcwd(), ".env"), help="Path to .env file")
    args = parser.parse_args()

    env = load_env(args.env_file)
    api_url = env.get("SITELINE_API_URL") or os.environ.get("SITELINE_API_URL")
    api_token = env.get("SITELINE_API_TOKEN") or os.environ.get("SITELINE_API_TOKEN")
    auth_header = env.get("SITELINE_AUTH_HEADER") or os.environ.get("SITELINE_AUTH_HEADER") or ""

    if not api_url or not api_token:
        print(
            "Missing SITELINE_API_URL and/or SITELINE_API_TOKEN. Set them in .env or env vars.",
            file=sys.stderr,
        )
        sys.exit(2)

    # Match backend behavior: backend expects full URL and uses it as-is.
    # Your .env uses /graphql already, but we keep it flexible.
    # If someone provides a base URL without /graphql, we append it.
    if not api_url.rstrip("/").endswith("/graphql"):
        api_url = api_url.rstrip("/") + "/graphql"

    for type_name in args.types:
        fields = list_fields(api_url, api_token, auth_header, type_name)
        print(f"--- {type_name} fields ({len(fields)}) ---")
        if not fields:
            print("(no fields or unknown type)")
        else:
            for f in fields:
                print(f"- {f}")
        print()


if __name__ == "__main__":
    main()

File: scripts/siteline_list_type_fields.js
/* eslint-disable no-console */
const fs = require('fs');
const path = require('path');
const process = require('process');
const fetch = require('node-fetch');

const SCHEMA_QUERY = `
query __TypeFields($name: String!) {
  __type(name: $name) {
    fields {
      name
    }
  }
}
`;

function loadEnv(envFile) {
  const env = {};
  if (!envFile || !fs.existsSync(envFile)) return env;

  const raw = fs.readFileSync(envFile, 'utf8');
  for (const lineRaw of raw.split(/\r?\n/)) {
    const line = lineRaw.trim();
    if (!line || line.startsWith('#')) continue;
    const eq = line.indexOf('=');
    if (eq === -1) continue;
    const k = line.slice(0, eq).trim();
    let v = line.slice(eq + 1).trim();
    // Remove wrapping quotes that are common in .env files
    v = v.replace(/^['"]/, '').replace(/['"]$/, '');
    env[k] = v;
  }
  return env;
}

function parseArgs(argv) {
  const out = { types: [], envFile: path.join(process.cwd(), '.env') };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--type') {
      const v = argv[i + 1];
      if (!v) throw new Error('Missing value after --type');
      out.types.push(v);
      i++;
      continue;
    }
    if (a === '--env-file') {
      const v = argv[i + 1];
      if (!v) throw new Error('Missing value after --env-file');
      out.envFile = v;
      i++;
      continue;
    }
    if (a === '--help' || a === '-h') {
      printHelpAndExit(0);
    }
  }
  if (!out.types.length) {
    printHelpAndExit(2);
  }
  return out;
}

function printHelpAndExit(code) {
  console.log('Usage: node scripts/siteline_list_type_fields.js --type Project --type Contract [--env-file /path/to/.env]');
  process.exit(code);
}

function getHeaders(apiToken, authHeader) {
  if (authHeader) return { [authHeader]: apiToken };
  return { Authorization: `Bearer ${apiToken}` };
}

async function listFields(apiUrl, apiToken, authHeader, gqlTypeName) {
  const headers = {
    ...getHeaders(apiToken, authHeader),
    'Content-Type': 'application/json',
  };

  const resp = await fetch(apiUrl, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      query: SCHEMA_QUERY,
      variables: { name: gqlTypeName },
    }),
  });

  if (!resp.ok) {
    const text = await resp.text().catch(() => '');
    throw new Error(`GraphQL HTTP ${resp.status}: ${text}`);
  }

  const payload = await resp.json();
  if (payload.errors && payload.errors.length) {
    throw new Error(`GraphQL errors: ${JSON.stringify(payload.errors)}`);
  }

  const t = payload?.data?.__type;
  if (!t || !t.fields) return [];

  return [...t.fields.map((f) => f.name).filter(Boolean)].sort();
}

async function main() {
  const args = parseArgs(process.argv);
  const env = loadEnv(args.envFile);

  const apiUrlRaw = env.SITELINE_API_URL || process.env.SITELINE_API_URL;
  const apiToken = env.SITELINE_API_TOKEN || process.env.SITELINE_API_TOKEN;
  const authHeader = (env.SITELINE_AUTH_HEADER || process.env.SITELINE_AUTH_HEADER || '').trim();

  if (!apiUrlRaw || !apiToken) {
    console.error('Missing SITELINE_API_URL and/or SITELINE_API_TOKEN. Set them in .env.');
    process.exit(2);
  }

  // Backend expects full URL. If user gives base URL, append /graphql.
  let apiUrl = apiUrlRaw.replace(/\/$/, '');
  if (!apiUrl.endsWith('/graphql')) apiUrl = `${apiUrl}/graphql`;

  for (const typeName of args.types) {
    const fields = await listFields(apiUrl, apiToken, authHeader, typeName);
    console.log(`--- ${typeName} fields (${fields.length}) ---`);
    if (!fields.length) {
      console.log('(no fields or unknown type)');
    } else {
      for (const f of fields) console.log(`- ${f}`);
    }
    console.log('');
  }
}

main().catch((err) => {
  console.error(err?.message ?? String(err));
  process.exit(1);
});

